### PR TITLE
docs: Add coding_guidelines.md

### PR DESCRIPTION
**Pull request by danobi**
Initial document. If anything in the document warrants debate, let's
do it here now.

##### Checklist

- [ ] Language changes are updated in `man/adoc/bpftrace.adoc` and if needed in `docs/reference_guide.md`
- [ ] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [ ] The new behaviour is covered by tests

### CODE REVIEW

File: docs/coding_guidelines.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+If something is not called out in this document, defer to [the C++ core
+guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines).
+In other words, this document supersedes the core guidelines.
+
+Note that the checked in code may not adhere to these guidelines yet.
+This document is intended to be referenced for new code. Ideally we also
+modify checked-in code to follow these guidelines (if/when there is time).
+
+## Error handling and exceptions
```
Comment (jordalgo) [inline docs/coding_guidelines.md:20]: Nit: would be good to include a small example.

File: docs/developers.md [hunk 2]
```diff
@@ -109,6 +109,14 @@ list of known such tests:
 What usually helps, is restarting the CI. This is simple on your own fork but
 requires one of the maintainers for pull requests.
 
+## Coding guidelines
+
+This is not about what the source code looks like. Rather, it's about the
```
Comment (jordalgo) [inline docs/developers.md:114]:
```suggestion
This is not about the format of the source code (we have clang-format). Rather, it's about the
```

File: docs/coding_guidelines.md (hunk 1 above)
Comment (danobi) [reply docs/coding_guidelines.md:20]: Yeah, good idea

File: docs/coding_guidelines.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+If an error is **not** recoverable, prefer throwing an exception. Broadly
+speaking, if you want to immediately terminate and show a message to the
+user, use an exception. Exceptions **should not** be used for recoverable
+errors.
+
+### Examples
+
+An example of a recoverable error would be failure to write an entry into
+a map. There are a number of reasons why writing to a map could fail
+(map is full, insufficient privileges, map does not exist yet, etc.) and
```
Comment (jordalgo) [inline docs/coding_guidelines.md:41]:
```suggestion
(map is full, insufficient privileges, map does not exist yet, etc...) and
```

File: docs/coding_guidelines.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+we can't know for sure at the point of failure what the reason is. So
+we need to propagate the error.
+
+An example of an unrecoverable error would be if `debugfs` not mounted.
```
Comment (jordalgo) [inline docs/coding_guidelines.md:45]: nit

```suggestion
An example of an unrecoverable error would be if `debugfs` failed to mount.
```

File: docs/coding_guidelines.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+speaking, if you want to immediately terminate and show a message to the
+user, use an exception. Exceptions **should not** be used for recoverable
+errors.
+
+### Examples
+
+An example of a recoverable error would be failure to write an entry into
+a map. There are a number of reasons why writing to a map could fail
+(map is full, insufficient privileges, map does not exist yet, etc.) and
+we can't know for sure at the point of failure what the reason is. So
```
Comment (jordalgo) [inline docs/coding_guidelines.md:42]:
```suggestion
we can't know for sure at the point of failure what the reason is. Additionally, this type of error might not be a reason fail the whole program. So
```

**Review by jordalgo — APPROVED** [review]
Thanks for getting this started! Two more minor nits but looks good otherwise.

**Review by ajor — APPROVED** [review]
Looks good, thanks for starting this!

File: docs/coding_guidelines.md (hunk 4 above)
Comment (danobi) [reply docs/coding_guidelines.md:45]: Should probably be "is not mounted". I don't believe we try to mount debugfs in our code (nor should we)

**Comment by danobi** [comment]
@viktormalik would like to get a look from you

File: docs/coding_guidelines.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+If something is not called out in this document, defer to [the C++ core
+guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines).
+In other words, this document supersedes the core guidelines.
+
+Note that the checked in code may not adhere to these guidelines yet.
+This document is intended to be referenced for new code. Ideally we also
+modify checked-in code to follow these guidelines (if/when there is time).
+
+## Error handling and exceptions
```
Comment (viktormalik) [inline docs/coding_guidelines.md:20]: Should we also mention how and when to use the `LOG(ERROR)` facility?

**Review by viktormalik — APPROVED** [review]
LGTM, just one remark but I'm ok adding that in another PR.

File: docs/coding_guidelines.md (hunk 6 above)
Comment (danobi) [reply docs/coding_guidelines.md:20]: Do you mean in context of AST passes?

Comment (viktormalik) [reply docs/coding_guidelines.md:20]: Well, we're using it in other places, too. But yes, in context of AST passes it's especially useful as using it makes the pass fail.

**Pull request merged**